File: 1999/1/6-1.py
#!/usr/bin/env python

import sys

def moveRight(grid, r, c, dimR, dimC, pathFound):
    grid[r][c] = 'A'
    if ((r == dimR - 1) and (c == dimC - 1)):
        pathFound = 1
        dumpGrid(grid, dimR, dimC)
    if (isValid(grid, r + 1, c, dimR, dimC)):
        pathFound = pathFound or moveDown(grid, r + 1, c, dimR, dimC, pathFound)
    if (not pathFound and isValid(grid, r, c + 1, dimR, dimC)):
        pathFound = pathFound or moveRight(grid, r, c + 1, dimR, dimC, pathFound)
    if (not pathFound and isValid(grid, r - 1, c, dimR, dimC)):
        pathFound = pathFound or moveUp(grid, r - 1, c, dimR, dimC, pathFound)
    grid[r][c] = 'Y'
    return pathFound

def moveLeft(grid, r, c, dimR, dimC, pathFound):
    grid[r][c] = 'A'
    if ((r == dimR - 1) and (c == dimC - 1)):
        pathFound = 1
        dumpGrid(grid, dimR, dimC)
    if (isValid(grid, r - 1, c, dimR, dimC)):
        pathFound = pathFound or moveUp(grid, r - 1, c, dimR, dimC, pathFound)
    if (not pathFound and isValid(grid, r, c - 1, dimR, dimC)):
        pathFound = pathFound or moveLeft(grid, r, c - 1, dimR, dimC, pathFound)
    if (not pathFound and isValid(grid, r + 1, c, dimR, dimC)):
        pathFound = pathFound or moveDown(grid, r + 1, c, dimR, dimC, pathFound)
    grid[r][c] = 'Y'
    return pathFound

def moveDown(grid, r, c, dimR, dimC, pathFound):
    grid[r][c] = 'A'
    if ((r == dimR - 1) and (c == dimC - 1)):
        pathFound = 1
        dumpGrid(grid, dimR, dimC)
    if (isValid(grid, r, c - 1, dimR, dimC)):
        pathFound = pathFound or moveLeft(grid, r, c - 1, dimR, dimC, pathFound)
    if (not pathFound and isValid(grid, r + 1, c, dimR, dimC)):
        pathFound = pathFound or moveDown(grid, r + 1, c, dimR, dimC, pathFound)
    if (not pathFound and isValid(grid, r, c + 1, dimR, dimC)):
        pathFound = pathFound or moveRight(grid, r, c + 1, dimR, dimC, pathFound)
    grid[r][c] = 'Y'
    return pathFound

def moveUp(grid, r, c, dimR, dimC, pathFound):
    grid[r][c] = 'A'
    if ((r == dimR - 1) and (c == dimC - 1)):
        pathFound = 1
        dumpGrid(grid, dimR, dimC)
    if (isValid(grid, r, c + 1, dimR, dimC)):
        pathFound = pathFound or moveRight(grid, r, c + 1, dimR, dimC, pathFound)
    if (not pathFound and isValid(grid, r - 1, c, dimR, dimC)):
        pathFound = pathFound or moveUp(grid, r - 1, c, dimR, dimC, pathFound)
    if (not pathFound and isValid(grid, r, c - 1, dimR, dimC)):
        pathFound = pathFound or moveLeft(grid, r, c - 1, dimR, dimC, pathFound)
    grid[r][c] = 'Y'
    return pathFound

def isValid(grid, r, c, dimR, dimC):
    return (r > -1) and (r < dimR) and (c > -1) and (c < dimC) and (grid[r][c] == 'Y')

def dumpGrid(grid, dimR, dimC):
    for i in xrange(dimR):
        print "  ",
        for j in xrange(dimC):
            if grid[i][j] == 'X' or grid[i][j] == 'Y':
                print "-",
            else:
                print "0",
        print

cnt = 1
while 1:
    line = sys.stdin.readline()
    if line == "":
        break
    dimR, dimC = line.split()
    dimR = int(dimR)
    dimC = int(dimC)
    grid = [0] * dimR
    for i in xrange(dimR):
        line = sys.stdin.readline()
        ln = [0] * dimC
        for j in xrange(dimC):
            ln[j] = line[j]
        grid[i] = ln
    print
    print "Maze %s:" % cnt
    pathFound = moveRight(grid, 0, 0, dimR, dimC, 0)
    if not pathFound:
        print "   No path exists."
    cnt += 1

File: 1999/1/6-1.cpp
#include <iostream>
#include <fstream>
#include <vector>
using namespace std;

void moveRight(vector< vector<char> >&, int, int);
void moveLeft(vector< vector<char> >&, int, int);
void moveDown(vector< vector<char> >&, int, int);
void moveUp(vector< vector<char> >&, int, int);

bool pathFound;
int dimC, dimR;

bool isValid(vector< vector<char> >& grid, int r, int c)
{
    return ((r > -1) && (r < dimR) && (c > -1) && (c < dimC) && (grid[r][c] == 'Y'));
}

void dumpGrid(vector< vector<char> >& grid)
{
    for (int i = 0; i < dimR; i++)
    {
        cout << "   ";
        for (int j = 0; j < dimC; j++)
        {
            if ((grid[i][j] == 'X') || (grid[i][j] == 'Y'))
                cout << "-";
            else
                cout << "0";
        }
        cout << endl;
    }
}

int main()
{
    ifstream in("prob1_in.txt");
    int cnt = 1;
    while (in >> dimR >> dimC)
    {
        vector< vector<char> > grid(dimR, vector<char>(dimC));
        pathFound = false;
        for (int i = 0; i < dimR; i++)
            for (int j = 0; j < dimC; j++)
                in >> grid[i][j];
        cout << "\nMaze " << cnt << ":" << endl;
        moveRight(grid, 0, 0);
        if (!pathFound)
            cout << "   No path exists." << endl;
        cnt++;
    }
}

void moveRight(vector< vector<char> >& grid, int r, int c)
{
    grid[r][c] = 'A';
    if ((r == dimR - 1) && (c == dimC - 1))
    {
        pathFound = true;
        dumpGrid(grid);
    }
    if (isValid(grid, r + 1, c))
        moveDown(grid, r + 1, c);
    if (!pathFound && isValid(grid, r, c + 1))
        moveRight(grid, r, c + 1);
    if (!pathFound && isValid(grid, r - 1, c))
        moveUp(grid, r - 1, c);
    grid[r][c] = 'Y';
}

void moveLeft(vector< vector<char> >& grid, int r, int c)
{
    grid[r][c] = 'A';
    if ((r == dimR - 1) && (c == dimC - 1))
    {
        pathFound = true;
        dumpGrid(grid);
    }
    if (isValid(grid, r - 1, c))
        moveUp(grid, r - 1, c);
    if (!pathFound && isValid(grid, r, c - 1))
        moveLeft(grid, r, c - 1);
    if (!pathFound && isValid(grid, r + 1, c))
        moveDown(grid, r + 1, c);
    grid[r][c] = 'Y';
}

void moveDown(vector< vector<char> >& grid, int r, int c)
{
    grid[r][c] = 'A';
    if ((r == dimR - 1) && (c == dimC - 1))
    {
        pathFound = true;
        dumpGrid(grid);
    }
    if (isValid(grid, r, c - 1))
        moveLeft(grid, r, c - 1);
    if (!pathFound && isValid(grid, r + 1, c))
        moveDown(grid, r + 1, c);
    if (!pathFound && isValid(grid, r, c + 1))
        moveRight(grid, r, c + 1);
    grid[r][c] = 'Y';
}

void moveUp(vector< vector<char> >& grid, int r, int c)
{
    grid[r][c] = 'A';
    if ((r == dimR - 1) && (c == dimC - 1))
    {
        pathFound = true;
        dumpGrid(grid);
    }
    if (isValid(grid, r, c + 1))
        moveRight(grid, r, c + 1);
    if (!pathFound && isValid(grid, r - 1, c))
        moveUp(grid, r - 1, c);
    if (!pathFound && isValid(grid, r, c - 1))
        moveLeft(grid, r, c - 1);
    grid[r][c] = 'Y';
}
